op-chain-ops/devkeys/devkeys.go

```diff
@@ -0,0 +1,125 @@
+)
+
+func (d Domain) String() string {
+	switch d {
+	case DefaultKeyDomain:
+		return "default"
+	case OperatorKeyDomain:
+		return "operator"
+	case UserKeyDomain:
+		return "user"
+	case SuperchainKeyDomain:
+		return "superchain"
+	default:
+		return fmt.Sprintf("unknown-domain-%d", uint64(d))
+	}
+}
+
+// Role separates the usage of development keys by exact role, within a Domain.
+// Some roles may not be applicable to every Domain, those can be ignored.
+type Role uint64
+
+const (
+	// DeployerRole is the deployer of contracts
+	DeployerRole Role = 0
+	// ProposerRole is the key used by op-proposer
+	ProposerRole Role = 1
+	// BatcherRole is the key used by op-batcher
+	BatcherRole Role = 2
+	// SequencerP2PRole is the key used to publish sequenced L2 blocks
+	SequencerP2PRole Role = 3
+	// ChallengerRole is the key used by op-challenger
+	ChallengerRole Role = 4
+	// L2ProxyAdminOwnerRole is the key that controls the ProxyAdmin predeploy in L2
+	L2ProxyAdminOwnerRole Role = 5
+	// SuperchainConfigGuardianRole is the Guardian of the SuperchainConfig
+	SuperchainConfigGuardianRole Role = 6
+	// L1ProxyAdminOwnerRole is the key that owns the ProxyAdmin on the L1 side of the deployment.
+	// This can be the ProxyAdmin of a L2 chain deployment, or a superchain deployment, depending on the domain.
+	L1ProxyAdminOwnerRole Role = 7
+	// BaseFeeVaultRecipientRole is the key that receives from the BaseFeeVault predeploy
+	BaseFeeVaultRecipientRole Role = 8
+	// L1FeeVaultRecipientRole is the key that receives from the L1FeeVault predeploy
+	L1FeeVaultRecipientRole Role = 9
+	// SequencerFeeVaultRecipientRole is the key that receives form the SequencerFeeVault predeploy
+	SequencerFeeVaultRecipientRole Role = 10
+	// DependencySetManagerRole is the key used to manage the dependency set of a superchain.
+	DependencySetManagerRole Role = 11
+)
+
+func (d Role) String() string {
+	switch d {
+	case DeployerRole:
+		return "deployer"
+	case ProposerRole:
+		return "proposer"
+	case BatcherRole:
+		return "batcher"
+	case SequencerP2PRole:
+		return "sequencer-p2p"
+	case ChallengerRole:
+		return "challenger"
+	case L2ProxyAdminOwnerRole:
+		return "l2-proxy-admin-owner"
+	case SuperchainConfigGuardianRole:
+		return "superchain-config-guardian"
+	case L1ProxyAdminOwnerRole:
+		return "l1-proxy-admin-owner"
+	case BaseFeeVaultRecipientRole:
+		return "base-fee-vault-recipient"
+	case L1FeeVaultRecipientRole:
+		return "l1-fee-vault-recipient"
+	case SequencerFeeVaultRecipientRole:
+		return "sequencer-fee-vault-recipient"
+	case DependencySetManagerRole:
+		return "dependency-set-manager"
+	default:
+		return fmt.Sprintf("unknown-role-%d", uint64(d))
+	}
+}
+
+// DevSecrets selects a secret-key based on domain, chain and role.
+// This is meant for dev-purposes only.
+// Secret keys should not directly be exposed to live production services.
+type DevSecrets interface {
+	Secret(domain Domain, chain *big.Int, role Role) (*ecdsa.PrivateKey, error)
```

Comment:
To be honest I don't think we need the `Domain`, it isn't super clear what it is used for. We don't necessarily need the role/domain combos because each role is generally used in a single domain. I think that chainid + role is good enough and removes extra cognitive overhead. Are you thinking that domain/role can mix together in some way to give extra flexibility?

Reply:
The main things I am hoping to separate:
- user accounts: I want e.g. 20 user accounts, and they shouldn't possibly overlap with the operator accounts. Tests have failed before in surprising ways because of accidental collisions between the usage of these.
- superchain vs regular chain: the superchain has its own `ProxyAdmin` deployment on L1, not part of any L2, but the exact same type of contract. The owner of that contract should have the same "role", but isn't the same domain purpose.
- future compatibility: if we add new kinds of things, we can do so, without breaking the existing usage.

So I do think the "domain" is useful to have there.

Reply:
I see your point, the thing that I feel slightly confused about is `Domain(OperatorKeyDomain) || Role(L2ProxyAdminOwnerRole)` vs `Domain(UserKeyDomain) || Role(L2ProxyAdminOwnerRole)`  - would it make sense to create an abstraction that allows these to both be expressed, since really only 1 should be used? Or am I misunderstanding how it should be used

Reply:
For some domains not all roles, or any, apply. But then we end up with more different types, not sure if it's worth it.

Reply:
Possible solution: remove the String function, and add a different one conditional on the Domain, and then not opinionate the role index globally per domain.

Reply:
Or better more native Go typing: combine them into a single Interface argument, that derives the HD-path, given a chain ID. And then we just have a user-account type by index, and an operator-account by enum.

Reply:
Implemented that approach, lmk what you think